src/main/kotlin/com/manimdsl/frontend/Type.kt

```diff
@@ -0,0 +1,38 @@
+package com.manimdsl.frontend
+
+
+sealed class Type: ASTNode()
+
+sealed class PrimitiveType: Type()
+object NumberType: PrimitiveType()
+
+sealed class DataStructureType(open var internalType: Type, open val methods: HashMap<String, DataStructureMethod>): Type() {
+    abstract fun containsMethod(method: String): Boolean
+    abstract fun getMethodByName(method: String): DataStructureMethod
+}
+
+// Pre-defined methods for each data structure
+open class DataStructureMethod(open val returnType: Type, open var argumentTypes: List<Type>)
+data class ErrorMethod(override val returnType: Type = NoType, override var argumentTypes: List<Type> = listOf()) : DataStructureMethod(returnType, argumentTypes)
+
+data class StackType(override var internalType: Type = NumberType,
+                     override val methods: HashMap<String, DataStructureMethod> = hashMapOf("push" to PushMethod(argumentTypes=listOf(NumberType)), "pop" to PopMethod(internalType))): DataStructureType(internalType, methods) {
```

Comment:
This file could get really long maybe consider creating a type package and moving this into a file called StackType.kt

Reply:
`Types` themselves are `ASTNodes` so I could move all of these files into an `ast` package instead?

Reply:
Also data classes can't inherit from sealed classes if they're placed in different files. We can change the sealed class to an open one if it gets too long but I'd prefer to keep them in the same file for now

Reply:
Yeah the sealed class point is a good one, we had this in WACC too, probably best to leave it in here